### DISCUSSION TITLE
STIR and optimising GPU support (i.e. parallelproj and NiftyPET)

### DISCUSSION THREAD

**Discussion by KrisThielemans** [UCL/STIR#1239] [Q&A]
STIR currently supports CUDA-enabled projectors from [NiftyPET](https://github.com/NiftyPET/NIPET/tree/master/niftypet/nipet) (for mMR only due to hard-wiring) and [Parallelproj](https://github.com/gschramm/parallelproj), but performance is suboptimal. We're thinking on how to speed this up. As I'm more familiar with parallelproj, we can start there. That STIR code was based on @rijobro's work on NiftyPET anyway.

The current STIR strategy is to call the GPU code for all projection data (and then sort things out on subsets afterwards. This is sub-optimal as well, but let's concentrate on how to speed the projection up first.)

Example is in the [forward projection](https://github.com/UCL/STIR/blob/4408419d9ef49c8295814087564272eed5dab706/src/recon_buildblock/Parallelproj_projector/ForwardProjectorByBinParallelproj.cxx#L148). Current strategy:
1. copy STIR array to `std::vector` (to get continuous memory) https://github.com/UCL/STIR/blob/4408419d9ef49c8295814087564272eed5dab706/src/recon_buildblock/Parallelproj_projector/ForwardProjectorByBinParallelproj.cxx#L148-L149
2. copy `std::vector` data to all GPUs https://github.com/UCL/STIR/blob/4408419d9ef49c8295814087564272eed5dab706/src/recon_buildblock/Parallelproj_projector/ForwardProjectorByBinParallelproj.cxx#L169-L170
3.  call parallelproj with image in the GPU memory, giving a pointer to data (stored contiguously by `ProjDataInMemory`, i.e. CPU). Note that `parallelproj` will do the [copy of projection data from GPU to host](https://github.com/gschramm/parallelproj/blob/618c63213f5c52463a9ef1abcb065db02cd00c2b/cuda/src/joseph3d_fwd_cuda.cu#L110) https://github.com/UCL/STIR/blob/4408419d9ef49c8295814087564272eed5dab706/src/recon_buildblock/Parallelproj_projector/ForwardProjectorByBinParallelproj.cxx#L181-L189. 
4. copy projection data from `ProjDataInMemory` to `ProjData` object that is actually given by the user (which could be in memory, but could also sit on disk, as `set_viewgram` etc are overloaded accordingly)

Some reasons for this are:
- step 1&2: `stir::Array` is currently not contiguously stored. This is being addressed in #1236, which means that step 1 could be avoided.
- step 3: GPU memory might be too small to store complete projection data, so you can ask `parallelproj` to do this in chunks.
- step 4: generally STIR is written to be able to handle projection data of any size, including LAFOV PET, it doesn't even need to fit in CPU memory. However, this is currently broken by the above strategy (and [in the sensitivity calculation as well](https://github.com/UCL/STIR/blob/4408419d9ef49c8295814087564272eed5dab706/src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx#L721-L723)).

Similar steps (in reverse) happen in the [back-projection](https://github.com/UCL/STIR/blob/4408419d9ef49c8295814087564272eed5dab706/src/recon_buildblock/Parallelproj_projector/BackProjectorByBinParallelproj.cxx#L133)

Some points from a discussion held with @casperdcl @markus-jehl @evgueni-ovtchinnikov and others on 31Aug2023:
- @casperdcl suggests that using `cudaMallocManaged` in a few places might avoid the explicit transfer (of images only?) between CPU and GPU (CUDA will take care of it). One option could be to use `cudaMallocManaged` for all `Array`s. This could be extended to numerical operations by either using libraries, or by using `#ifdef`s. Example of this is in https://github.com/AMYPAD/NumCu/blob/main/numcu/src/elemwise.cu
- @casperdcl wrote [CuVec](https://github.com/AMYPAD/CuVec) to be able to auto-magically transport vectors between CUDA/C/Python (via SWIG), such that other libraries in Python can be used to do the operations on the arrays. This could be useful for STIR as well.

Some anticipated difficulties:
- `cudaMallocManaged` will probably fail when asking for more memory than available on the GPU. I don't know if this is a problem when allocating multiple blocks (i.e. do they all need to fit in GPU together?)
- projection data can be very big (Siemens Vision sinograms are 2.8GB, LAFOV data will be a lot bigger). Avoiding step 4 means taking care of the "chunking" directly without going via `ProjDataInMemory`, but this needs a refactor of the `ForwardProjectorByBin` class.
- without doing timings on these various steps, we don't know which ones we should address first. For example,  
    - according to some timings in #1237 copying an image (step 1) seems to take only ~2 ms.
    - #1236 seems to make no difference in timings of current STIR (seee discussion there)

Comments/suggestions/PRs welcome!

**Comment by KrisThielemans** [comment]
@casperdcl, just replacing the `std::vector` in step 1 with a `cudaMalloc`ed  memory block (or `CuVec` vector) is presumably not going to make any difference at all, as it will need to upload the data to GPU anyway. Correct? (By the way, when using multiple GPUs, does a "shared" array get copied to automatically to whatever GPU needs the data as well?)

**Reply by casperdcl** [reply]
I haven't stress tested multi-GPU behaviour, but happy to do so. TL;DR calling `cudaSetDevice(int)` or `cuvec.dev_set(int)` should be enough :)

**Comment by KrisThielemans** [comment]
Note that being able to expose the GPU arrays directly to Python would presumably speed-up `pytorch` and `tensorflow` a lot, although possibly that needs a lot more work, which might need a different Discussion in that case.

**Reply by casperdcl** [reply]
`CuVec` python objects [can be directly understood](https://amypad.github.io/CuVec/#external-projects) by `numpy`, `tensorflow`, `pytorch`, `cupy`, etc (i.e. all these common libs will work without memcopies).

**Reply by KrisThielemans** [reply]
That's great! We'd need to get our head round how to create a "rich" object (with all the meta-info AND the data) that can be used transparently, or maybe that can't be done.

**Reply by KrisThielemans** [reply]
Reading up a bit, this is done by implementing the `__cuda_array_interface__` property.

OpenRTK has a PR to https://github.com/RTKConsortium/ITKCudaCommon/pull/43 to convert an itk.CudaImage to, e.g., pytorch tensor or cupy array, without GPU to CPU transfers. The magic seems to sit in the SWIG file https://github.com/RTKConsortium/ITKCudaCommon/pull/43/files#diff-b08959cf69284340900976ab6e14570a65e78745850a4f8de0942d908ba5a02d
```
%extend itkCudaImage@CudaImageTypes@{
    %pythoncode %{
    @property
    def __cuda_array_interface__(self):
   ...
```

**Comment by casperdcl** [comment]
@gschramm working on v2.0.0 release of https://github.com/KUL-recon-lab/libparallelproj:

- all pointers can be host, CUDA device, or CUDA managed
- docs for C-API: https://libparallelproj.readthedocs.io/en/latest/c_api.html
- non-TOF & TOF **sinogram** projector interfaces have not changed
- **listmode** projectors have an additional `num_tofbins` argument (breaking change)

**Reply by gschramm** [reply]
just a small heads up. v2.0.0 is ready to be released. I am just optimizing the build of the cuda/non-cuda versions + optional minimal python interface for conda-forge using the new "rattler-build" tool + the new recipe.yaml format (both awesome :)

**Comment by KrisThielemans** [comment]
A lot of progress was made during and after the SyneRBI AI-RBI hackathon. See https://github.com/SyneRBI/SIRF-Contribs/discussions/31#discussioncomment-15981589

**Comment by KrisThielemans** [comment]
There seem to be a few different approaches to get STIR images use cuda managed pointers. One of which is to add an "allocator" argument to `stir::Array` and use that, but that probably needs a fair amount of surgery.

However, `stir::Array` can now accept a `shared_ptr` as part of its constructor, so simpler seems to be to let `DiscretisedDensity` use that. My impression is that this only needs changes in
https://github.com/UCL/STIR/blob/4368758556118fc59363403e30ef6f8e7f0be25d/src/include/stir/DiscretisedDensity.inl#L40-L53
where as opposed to
```c++
base_type(range_v)
```
we would use something like (not 100% sure! @casperdcl will correct me)
```c++
base_type(range_v, std::allocate_shared<T>(cuvec::CuAlloc(), range_v.size_all())
```
The disadvantage of `std::allocate_shared` is likely that the internal counter would be part of the Cuda managed memory, as noted by @SimonRit. That is likely sub-optimal. Better might be to use a `shared_ptr` that is constructed with its own destructor. This could be done via thin wrapper around `shared_ptr` as in https://github.com/lkoshale/cuda-smart-pointers/blob/33894ff88f3d8558f8845aec8bd33722feadd594/src/memory.cuh#L36-L39.

The approach of changing only the `DiscretisedDensity` constructor would need minimal code changes. However, any call to `resize` would fall back to "normal" allocation. I have no idea how often that happens.

**Reply by casperdcl** [reply]
Makes sense... happy to add shared/uniq support to CuVec in case it makes life easier (more permissive licence than the repo you linked).

**Reply by KrisThielemans** [reply]
I certainly think adding that repo would be a bad idea and therefore thought to do this "by-hand" for the moment. But if it's easy to add it to CuVec, that'd be great. Seems useful in more general situations.

**Comment by KrisThielemans** [comment]
Valerio and I discussed ways to hide complicated cuda fall-backs by introducing some additional functions. After reading a bit more, I came up with the untested example
```c++
template < class T>
void
mycopy(T& to, const T& from)
{
   const auto begin_to = to.begin();
   const auto end_to = to.end();
   const auto begin_from = from.begin();

   if constexpr (std::is_trivially_copyable_v<decltype(*begin_to)>)
   {
      if (on_gpu)
       {
          cudaMemcpy(begin_to, begin_from, (end_to - begin_to) * sizeof(*begin_to));
       }
       else
       {
          std::memcpy(begin_to, begin_from, (end_to - begin_to) * sizeof(*begin_to));
       }
    }
    else
    {
        std::copy(begin_to, end_to, begin_from);
    }
}
```
Some notes:
- we don't need the explicit `memcpy` as the compiler would do that for us, see #1259, but this might be easier to understand.
- This usage looks very much like [std::ranges::copy](https://en.cppreference.com/cpp/algorithm/ranges/copy) except for the order of arguments which isn't great (I followed STIR convention of "destination first").
- Would have to think about the return argument.
- Actual `std::copy` normally falls back to `memmove`, which is safer, but AFAIK only for overlapping ranges, which we wouldn't have.
- Note added 23/7/2026: will need to check if arrays are contiguous. e.g. `to.size()` compare to `end_to - begin_to`
